tests/kernel/smp/src/main.c

```diff
@@ -37,45 +52,354 @@ void t2_fn(void *a, void *b, void *c)
+ * will get scheduled in uniprocessor mode.
+ *
+ * For ESP32 only (has only 2 cores): Since t2 thread has
+ * higher priority than test_main and test_main is preemptive
+ * test_main should be moved to timeout queue and t2
+ * thread should get scheduled instead. So t2 thread should
+ * be in one core and test_smp_coop_threads should be
+ * another core.
+ */
+void test_smp_coop_threads(void)
+{
+	/* In a multi processor mode, this thread should have
+	 * been scheduled on another core other than the one
+	 * which has test_main scheduled
 	 */
-	k_thread_create(&t2, t2_stack, T2_STACK_SIZE, t2_fn,
-			NULL, NULL, NULL,
-			CONFIG_MAIN_THREAD_PRIORITY + 1, 0, K_NO_WAIT);
+	int i;
+
+	k_tid_t tid = k_thread_create(&t2, t2_stack, STACK_SIZE, t2_fn,
+				      NULL, NULL, NULL,
+				      K_PRIO_COOP(1), 0, K_NO_WAIT);
 
 	for (i = 0; i < 10; i++) {
-		/* Wait slightly longer than the other thread so our
-		 * count will always be lower
-		 */
 		k_busy_wait(DELAY_US + (DELAY_US / 8));
 
-		if (t2_count <= i) {
-			ok = 0;
-			break;
-		}
+		zassert_true(t2_count > i, "Another thread isn't"
+			     " scheduled to run on other core");
+	}
+	k_thread_abort(tid);
+}
+
+static void smp_thread_entry(void *p1, void *p2, void *p3)
+{
+	int tnum = (int)p1;
+
+	TC_PRINT("Thread started\n");
+	k_busy_wait(THREAD_DELAY);
+
+	tdata[tnum].executed = 1;
+	TC_PRINT("Thread execution complete\n");
+}
+
+/**
+ * @brief Verify round robin on one core with coop thread
+ *
+ * @details The cooperative test creates THREAD_NUM preemptive
+ * threads of same priority and checks the round robin behavior.
+ * the cooperative thread will be blocking one core hence
+ * preemptive threads should run on another core in round robin.
+ */
+static void test_smp_single_core_coop_rr_threads(void)
+{
+	int i;
+
+	k_sched_time_slice_set(100, 0);
+
+	for (i = 0; i < THREADS_NUM; i++) {
+		tdata[i].executed = 0;
+		tdata[i].tid = k_thread_create(&tthread[i], tstack[i],
+					       STACK_SIZE, smp_thread_entry,
+					       (void *)i, NULL, NULL,
+					       K_PRIO_PREEMPT(
+						       K_LOWEST_THREAD_PRIO - 1),
+					       0, 0);
+	}
+
+	k_busy_wait(THREAD_DELAY * THREADS_NUM + 10);
+
+	for (i = 0; i < THREADS_NUM; i++) {
+		k_thread_abort(tdata[i].tid);
+	}
+	/* Test point: All THREADS_NUM spawned threads must have
+	 * executed in another core
+	 */
+	for (i = 0; i < THREADS_NUM; i++) {
+		zassert_true(tdata[i].executed == 1, "Threads didn't get"
+			     " scheduled on another core");
+	}
+}
+
+/**
+ * @brief Verify round robin on one core with preempt thread
+ *
+ * @details The preemptive thread of higher priority creates
+ * lower priority threads. So higher priority thread should occupy
+ * one core and lower priority threads should run in round
+ * robin.
+ */
+static void test_smp_single_core_preemp_rr_threads(void)
+{
+	int i;
+
+	k_thread_priority_set(k_current_get(),
+			      K_PRIO_PREEMPT(K_HIGHEST_THREAD_PRIO));
```

Comment:
This bit is wrong.  K_HIGHEST_THREAD_PRIORITY is **not** going to be a preemptive thread unless CONFIG_NUM_COOP_PRIORITIES is zero (and to be honest I sorta suspect things would break if you tried that), so I don't understand the meaning here.